polaris/ocean/convergence/spherical/analysis.py

```diff
@@ -0,0 +1,332 @@
+                                          time=eval_time * s_per_day,
+                                          zidx=zidx)
+        field_mpas = ds_out[variable_name].values
+        diff = field_exact - field_mpas
+
+        if error_type == 'l2':
+            area_cell = ds_mesh.areaCell.values
+            total_area = np.sum(area_cell)
+            den_l2 = np.sum(field_exact**2 * area_cell) / total_area
+            num_l2 = np.sum(diff**2 * area_cell) / total_area
+            error = np.sqrt(num_l2) / np.sqrt(den_l2)
+        elif error_type == 'inf':
+            error = np.amax(diff) / np.amax(np.abs(field_exact))
+        else:
+            raise ValueError(f'Unsupported error type {error_type}')
+
+        return error
+
+    def exact_solution(self, mesh_name, field_name, time, zidx=None):
+        """
+        Get the exact solution
+
+        Parameters
+        ----------
+        field_name : str
+            The name of the variable of which we evaluate convergence
+            For the default method, we use the same convergence rate for all
+            fields
+
+        time : float
+            The time at which to evaluate the exact solution in seconds.
+            For the default method, we always use the initial state.
+
+        Returns
+        -------
+        solution: np.ndarray of type float
+            The minimum convergence rate
+        """
+
+        ds_init = xr.open_dataset(f'{mesh_name}_init.nc')
+        ds_init = ds_init.isel(Time=0)
+        if zidx is not None:
+            ds_init = ds_init.isel(nVertLevels=zidx)
+
+        return ds_init[field_name]
+
+    def convergence_parameters(self, field_name=None):
+        """
+        Get convergence parameters
+
+        Parameters
+        ----------
+        field_name : str
+            The name of the variable of which we evaluate convergence
+            For the default method, we use the same convergence rate for all
+            fields
+
+        Returns
+        -------
+        conv_thresh: float
+            The minimum convergence rate
+
+        conv_thresh: float
+            The maximum convergence rate
+        """
+        config = self.config
+        section = config['spherical_convergence']
+        conv_thresh = section.getfloat('convergence_thresh')
+        conv_max = section.getfloat('convergence_max')
+        error_type = section.get('error_type')
+        return conv_thresh, conv_max, error_type
+
+
+def _time_index_from_xtime(xtime, dt_target):
+    """
+    Determine the time index at which to evaluate convergence
+
+    Parameters
+    ----------
+    xtime: list of str
+        Times in the dataset
+    dt_target: float
+        Time in seconds at which to evaluate convergence
+
+    Returns
+    -------
+    tidx: int
+        Index in xtime that is closest to dt_target
+    """
+    t0 = datetime.datetime.strptime(xtime[0].decode(),
+                                    '%Y-%m-%d_%H:%M:%S')
```

Comment:
It's great to know that this works!  I think I had had trouble with years lower than around 1600 in the past but maybe this has been fixed.  I'm also surprised that the `xtime` variable can be decoded with `decode()`.  This must be a recent improvement in `xarray`.

Reply:
Maybe we could have a `polaris.time` module and put that there?  It would be really handy to have something equivalent for converting from seconds to `xtime` values, and that could also go there in the future.